File: legacy/src/main/rs/emulate/legacy/archive/ArchiveEntry.java
package rs.emulate.legacy.archive;

import rs.emulate.shared.util.ByteBufferUtils;
import rs.emulate.shared.util.DataBuffer;

/**
 * A single entry in an {@link Archive}. This class is immutable.
 *
 * @author Graham
 * @author Major
 */
public final class ArchiveEntry {

	/**
	 * The buffer of this entry.
	 */
	private final DataBuffer buffer;

	/**
	 * The identifier of this entry.
	 */
	private final int identifier;

	/**
	 * Creates the archive entry.
	 *
	 * @param identifier The identifier.
	 * @param buffer The buffer containing this entry's data.
	 */
	public ArchiveEntry(int identifier, DataBuffer buffer) {
		this.identifier = identifier;
		this.buffer = buffer.asReadOnlyBuffer();
	}

	/**
	 * Creates the archive entry.
	 *
	 * @param name The name of the archive.
	 * @param buffer The buffer containing this entry's data.
	 */
	public ArchiveEntry(String name, DataBuffer buffer) {
		this(ArchiveUtils.hash(name), buffer);
	}

	@Override
	public boolean equals(Object object) {
		if (object instanceof ArchiveEntry) {
			ArchiveEntry other = (ArchiveEntry) object;
			return identifier == other.identifier && buffer.equals(other.buffer);
		}

		return false;
	}

	/**
	 * Gets a deep copy of the buffer of this entry. The returned buffer will not be read-only (see
	 * {@link ByteBufferUtils#copy}.
	 *
	 * @return The buffer of this entry.
	 */
	public DataBuffer getBuffer() {
		return buffer.copy();
	}

	/**
	 * Gets the identifier of this entry.
	 *
	 * @return The identifier of this entry.
	 */
	public int getIdentifier() {
		return identifier;
	}

	/**
	 * Gets the size of this entry (i.e. the capacity of the {@link DataBuffer} backing it), in bytes.
	 *
	 * @return The size of this entry.
	 */
	public int getSize() {
		return buffer.limit();
	}

	@Override
	public int hashCode() {
		return 31 * buffer.hashCode() + identifier;
	}

	/**
	 * Returns a new archive entry with this entry's identifier, but the contents as the specified {@link DataBuffer}.
	 *
	 * @param buffer The byte buffer.
	 * @return The new archive entry.
	 */
	public ArchiveEntry update(DataBuffer buffer) {
		return new ArchiveEntry(identifier, buffer);
	}

}

File: legacy/src/main/kotlin/rs/emulate/legacy/archive/ArchiveEntry.kt
package rs.emulate.legacy.archive

import io.netty.buffer.ByteBuf
import rs.emulate.legacy.archive.Archive.Companion.entryHash

/**
 * A single entry in an [Archive].
 *
 * @param identifier The identifier.
 * @param buffer The buffer containing this entry's data.
 */
class ArchiveEntry(val identifier: Int, buffer: ByteBuf) {

    init {
        require(buffer.readerIndex() == 0) { "Cannot create an ArchiveEntry with a partially-read buffer" }
    }

    /**
     * The buffer of this entry.
     */
    val buffer: ByteBuf = buffer.copy()
        get() = field.copy()

    /**
     * Gets the size of this entry (i.e. the capacity of the [ByteBuf] backing it), in bytes.
     */
    val size: Int
        get() = buffer.readableBytes()

    /**
     * Creates the archive entry.
     *
     * @param name The name of the archive.
     * @param buffer The buffer containing this entry's data.
     */
    constructor(name: String, buffer: ByteBuf) : this(name.entryHash(), buffer)

    override fun equals(other: Any?): Boolean {
        if (other is ArchiveEntry) {
            return identifier == other.identifier && buffer == other.buffer
        }

        return false
    }

    override fun hashCode(): Int {
        return 31 * buffer.hashCode() + identifier
    }

}
